packages/scm/src/browser/scm-navigable-list-widget.tsx

```diff
@@ -0,0 +1,120 @@
+/********************************************************************************
+ * Copyright (C) 2018 TypeFox and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { Message } from '@phosphor/messaging';
+import { inject, injectable } from 'inversify';
+import { Key, LabelProvider, ReactWidget } from '@theia/core/lib/browser';
+import URI from '@theia/core/lib/common/uri';
+import { ScmResourceGroup } from './scm-service';
+
+@injectable()
+export abstract class ScmNavigableListWidget<T extends { group: ScmResourceGroup, selected?: boolean, sourceUri: URI }> extends ReactWidget {
+
+    protected scmNodes: T[];
+    private _scrollContainer: string;
+
+    @inject(LabelProvider) protected readonly labelProvider: LabelProvider;
+
+    constructor() {
+        super();
+        this.node.tabIndex = 0;
+    }
+
+    protected onActivateRequest(msg: Message): void {
+        super.onActivateRequest(msg);
+        this.update();
+        this.node.focus();
+    }
+
+    protected set scrollContainer(id: string) {
+        this._scrollContainer = id + Date.now();
+    }
+
+    protected get scrollContainer(): string {
+        return this._scrollContainer;
+    }
+
+    protected addScmListNavigationKeyListeners(container: HTMLElement) {
```

Comment:
keep all return type explicit, it helps preventing surprising breaking changes latest, please review code for it

Reply:
done